File: src/extras.rs
//! Fetches supplementary album/artist metadata not captured by the qobuz-api models.

use std::time::{SystemTime, UNIX_EPOCH};

use qobuz_api::{api::{requests::build_url_with_params, service::QobuzApiService}, signing::sign_request};
use serde::Deserialize;

use crate::errors::AppError;

// ── Album extras ─────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct AlbumExtras {
    pub description: Option<String>,
    #[serde(default)]
    pub goodies: Option<Vec<Goody>>,
}

#[derive(Debug, Deserialize)]
pub struct Goody {
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub original_url: String,
    pub file_format_id: Option<u32>,
}

impl Goody {
    pub fn best_url(&self) -> &str {
        if !self.original_url.is_empty() {
            &self.original_url
        } else {
            &self.url
        }
    }

    pub fn is_pdf(&self) -> bool {
        self.url.ends_with(".pdf") || self.original_url.ends_with(".pdf")
            || self.file_format_id == Some(21)
    }
}

// ── Artist extras ─────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct ArtistExtras {
    pub biography: Option<Bio>,
    pub image: Option<ArtistImage>,
}

#[derive(Debug, Deserialize)]
pub struct Bio {
    pub text: Option<String>,
    pub summary: Option<String>,
}

impl Bio {
    pub fn best_text(&self) -> Option<&str> {
        self.text.as_deref().or(self.summary.as_deref())
    }
}

#[derive(Debug, Deserialize)]
pub struct ArtistImage {
    pub mega: Option<String>,
    #[serde(rename = "extralarge")]
    pub extra_large: Option<String>,
    pub large: Option<String>,
    pub medium: Option<String>,
}

impl ArtistImage {
    pub fn best_url(&self) -> Option<&str> {
        self.mega
            .as_deref()
            .or(self.extra_large.as_deref())
            .or(self.large.as_deref())
            .or(self.medium.as_deref())
    }
}

// ── Fetch helpers ─────────────────────────────────────────────────────────────

fn now_ts() -> String {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs()
        .to_string()
}

fn signed_url(base_url: &str, endpoint: &str, mut params: Vec<(String, String)>, app_id: &str, app_secret: &str) -> String {
    params.push(("app_id".to_string(), app_id.to_string()));
    params.push(("request_ts".to_string(), now_ts()));
    let sig = sign_request("GET", endpoint, &mut params, app_secret);
    params.push(("request_sig".to_string(), sig));
    build_url_with_params(base_url, endpoint, &params)
}

fn get_json<T: serde::de::DeserializeOwned>(url: &str, auth_token: &str) -> Result<T, AppError> {
    let resp = reqwest::blocking::Client::new()
        .get(url)
        .header("X-User-Auth-Token", auth_token)
        .send()
        .map_err(|e| AppError::Other(e.to_string()))?;

    resp.json::<T>().map_err(|e| AppError::Other(e.to_string()))
}

pub fn fetch_album_extras(service: &QobuzApiService, album_id: &str) -> Result<AlbumExtras, AppError> {
    let token = service.require_auth_token().map_err(|e| AppError::Other(e.to_string()))?;
    let url = signed_url(
        service.base_url(),
        "/album/get",
        vec![("album_id".to_string(), album_id.to_string())],
        &service.app_id,
        service.app_secret(),
    );
    get_json(&url, token)
}

pub fn fetch_artist_extras(service: &QobuzApiService, artist_id: i32) -> Result<ArtistExtras, AppError> {
    let token = service.require_auth_token().map_err(|e| AppError::Other(e.to_string()))?;
    let url = signed_url(
        service.base_url(),
        "/artist/get",
        vec![("artist_id".to_string(), artist_id.to_string())],
        &service.app_id,
        service.app_secret(),
    );
    get_json(&url, token)
}

File: src/extras.cpp
#include "extras.hh"
#include "i18n.hh"

#include <cstdio>
#include <filesystem>
#include <fstream>

#include <nlohmann/json.hpp>
#include <arc/http.hh>
#include <api/service.hh>
#include <api/requests.hh>

namespace fs = std::filesystem;

namespace extras {

// Download url to dest using the CDN client; skip if dest already exists.
static void download_file(const arc::HttpClient &client,
                          const std::string &url,
                          const fs::path &dest,
                          const char *label)
{
    if (fs::exists(dest)) return;
    auto res = client.download_to_file(url, dest.u8string());
    if (!res.ok())
        std::fprintf(stderr, "%s %s: %s\n", i18n::t("warning_could_not_download"),
                     label, res.error().message.c_str());
    else
        std::printf("%s %s\n", i18n::t("saved"), dest.u8string().c_str());
}

static void save_text(const std::string &text, const fs::path &dest) {
    if (fs::exists(dest)) return;
    std::ofstream f(dest);
    if (!f) {
        std::fprintf(stderr, "%s %s\n", i18n::t("warning_could_not_save"),
                     dest.u8string().c_str());
        return;
    }
    f << text;
    std::printf("%s %s\n", i18n::t("saved"), dest.u8string().c_str());
}

// Rewrite "…_600.jpg" → "…_org.jpg" to get full-resolution cover.
static std::string best_cover_url(const std::string &url) {
    auto pos = url.rfind("600");
    if (pos == std::string::npos) return url;
    return url.substr(0, pos) + "org" + url.substr(pos + 3);
}

void save_album_extras(const kb::QobuzApiService &svc,
                       const std::string &album_id,
                       const fs::path &album_dir)
{
    auto tok_r = svc.require_auth_token();
    if (!tok_r.ok()) return;
    auto auth = svc.request_auth(tok_r.value());

    kb::api::Params params = {{"album_id", album_id}};
    auto raw = kb::api::signed_get_raw(svc.http_client(), svc.base_url(),
                                       "/album/get", std::move(params), auth);
    if (!raw.ok()) {
        std::fprintf(stderr, "%s %s\n", i18n::t("warning_album_extras"),
                     raw.error().message.c_str());
        return;
    }

    auto j = nlohmann::json::parse(raw.value(), nullptr, false);
    if (j.is_discarded()) return;

    // Album description
    if (j.contains("description") && j["description"].is_string()) {
        std::string desc = j["description"].get<std::string>();
        if (!desc.empty())
            save_text(desc, album_dir / "album_description.txt");
    }

    // PDF booklet (goodies array)
    if (j.contains("goodies") && j["goodies"].is_array()) {
        for (auto &g : j["goodies"]) {
            std::string url_val = g.value("original_url", g.value("url", std::string{}));
            bool is_pdf = (g.value("file_format_id", 0) == 21)
                       || (!url_val.empty() && url_val.rfind(".pdf") == url_val.size() - 4);
            if (is_pdf && !url_val.empty()) {
                download_file(svc.cdn_client(), url_val,
                              album_dir / "booklet.pdf", "booklet");
                break;
            }
        }
    }
}

void save_artist_extras(const kb::QobuzApiService &svc,
                        int artist_id,
                        const fs::path &artist_dir)
{
    auto tok_r = svc.require_auth_token();
    if (!tok_r.ok()) return;
    auto auth = svc.request_auth(tok_r.value());

    kb::api::Params params = {{"artist_id", std::to_string(artist_id)}};
    auto raw = kb::api::signed_get_raw(svc.http_client(), svc.base_url(),
                                       "/artist/get", std::move(params), auth);
    if (!raw.ok()) {
        std::fprintf(stderr, "%s %s\n", i18n::t("warning_artist_extras"),
                     raw.error().message.c_str());
        return;
    }

    auto j = nlohmann::json::parse(raw.value(), nullptr, false);
    if (j.is_discarded()) return;

    // Artist bio
    if (j.contains("biography") && j["biography"].is_object()) {
        auto &bio = j["biography"];
        std::string text = bio.value("text", bio.value("summary", std::string{}));
        if (!text.empty())
            save_text(text, artist_dir / "artist_bio.txt");
    }

    // Artist image: try image.mega → extralarge → large → medium
    if (j.contains("image") && j["image"].is_object()) {
        auto &img = j["image"];
        for (const char *key : {"mega", "extralarge", "large", "medium"}) {
            if (img.contains(key) && img[key].is_string()) {
                std::string url_val = img[key].get<std::string>();
                if (!url_val.empty()) {
                    download_file(svc.cdn_client(), url_val,
                                  artist_dir / "artist.jpg", "artist image");
                    break;
                }
            }
        }
    }
}

} // namespace extras
